### ISSUE TITLE
InvokeTransaction V1 mismatch between docs.starknet.io and jsonrpc/models/codegen.rs

### ISSUE THREAD

**Issue by Eikix** [xJonathanLEI/starknet-rs#296]
Hi!

I'm going through the Starknet documentation to gather information about `Invoke Transaction V1`.

[Here is their doc](https://docs.starknet.io/documentation/architecture_and_concepts/Blocks/transactions/#invoke_transaction_version_1):

<img width="844" alt="image" src="https://user-images.githubusercontent.com/66871571/216660608-c8b46bf9-9dc1-4e36-8a47-e4ee19c68009.png">

Here is the signature in the Rust code:

```rust
/// Version 1 invoke transaction.
///
/// Initiates a transaction from a given account.
#[derive(Debug, Clone)]
pub struct InvokeTransactionV1 {
    /// The hash identifying the transaction
    pub transaction_hash: FieldElement,
    /// The maximal fee that can be charged for including the transaction
    pub max_fee: FieldElement,
    pub signature: Vec<FieldElement>,
    pub nonce: FieldElement,
    pub sender_address: FieldElement,
    /// The data expected by the account's `execute` function (in most usecases, this includes the
    /// called contract address and a function selector)
    pub calldata: Vec<FieldElement>,
}
```

What I notice is:
- There is a transaction_hash in starknet-rs but not in the starknet doc,
- There is a "version" field in Starknet docs, but I gather from the starknet-rs enum V1 that the value "version" is hardcoded to 1 when using the InvokeTransaction::V1.

Question: What is the `transaction_hash` field meant to be?

**Comment by Eikix** [comment]
My bad, it appears what is equivalent to an "unsubmitted" (yet) transactionV1 is `BroadcastedInvokeTransactionV1`! Will try playing around with it

**Comment by Eikix** [comment]
I notice that the `BroadcastedInvokeTransactionV1` type `Initiates a transaction from a given account.`

This means that `calldata` is actually what is passed into the `execute` entrypoint (apparently: `The data expected by the account's `execute` function (in most usecases, this includes the called contract address and a function selector`)

Does this mean that the `validate` is called some place else?



Here is my usecase:
I'm looking to submit a transaction to the sequencer with a sender_address pointing to a "homemade" account abstraction contract so that the StarknetOS calls `validate` on it
I'm not looking to provide a function selector for "execute"

Is the method `add_invoke_transaction` adapted?

**Comment by Eikix** [comment]
Nevermind!! I was having issues but apparently they were due to the fact that `FieldElement::MAX` is not a valid parameter for `max_fee` field

**Comment by xJonathanLEI** [comment]
Hi @Eikix. I agree that the naming of the raw value object types can be a bit confusing. To clarify:

- `starknet::core::types::InvokeFunctionTransaction` is the _confirmed_ form of the invoke transaction type, aka what's included in a block for example.
- If you're _constructing_ a transaction, `starknet::core::types::InvokeFunctionTransactionRequest` is what you're looking for.

The matter is further complicated as currently the lib is still sequencer-centric: all the low level types were modeled after the sequencer gateway types. The types like `BroadcastedInvokeTransactionV1` living under `starknet::providers::jsonrpc::models` are actually JSON-RPC specific. See #173 for more details.

However, even if you're only using JSON-PRC (and you should), you should probably still use the `core` types (not the ones living under `starknet::providers::jsorpc::models`). This is because it's the best to interface with the generic `Provider` trait, and the trait uses the `core` types. (It wasn't even possible to use `Provider` for JSON-RPC but that changed with #291)

And one more thing: you seem to be using low-level types directly but I actually wouldn't recommend that. If you're playing around with a custom account contract then the most idiomatic way of using it would be to implement the `Account` and `ConnectedAccount` traits:

```rust
struct MyAccount {
  ...
}

impl Account for MyAccount {
  ...
}

impl ConnectedAccount for MyAccount {
  ...
}
```

Basically just try to model after `SingleOwnerAccount` but with your own signature generation logic. Doing this was made simple thanks to #271, before which you'd need to write a lot of boilerplate code.

Hope this helps.

**Comment by xJonathanLEI** [comment]
Oh I just realized the code snippet you showed was actually from `jsonrpc::models`. My advice of switching to `core` types and thus `Provider` still holds.

Btw it's pretty much impossible for the JSON-RPC model types here to be different from the spec. They're [auto generated](https://github.com/xJonathanLEI/starknet-jsonrpc-codegen) exactly to avoid that issue.

**Comment by xJonathanLEI** [comment]
In any case I'm closing this optimistically as there seems to be no action required. Please feel free to re-open otherwise.

**Comment by xJonathanLEI** [comment]
Also I believe the docs you're referring to is for the sequencer gateway (legacy) while the snippet is for JSON-RPC, adding to the confusion even more.

**Comment by Eikix** [comment]
Hey:)! Thanks for the detailed answer and your time.

My understanding is that my transaction will have the following flow:
1. submit tx to starknet
2. starknetOS will call _validate_ on the address I provided in `InvokeTransactionRequest.contract_address`
3. if the account's _validate_ succeeds, _execute_ is called 

I'm not familiar with how the abstraction is constructed around Account

**Comment by Eikix** [comment]
What I still have trouble understanding is why `add_invoke_transaction` (my understanding is that it submits a transaction to Starknet via JSON RPC) takes as an argument a `BroadcastedInvokeTransaction` and not an `InvokeTransactionRequest`?

**Comment by xJonathanLEI** [comment]
> I'm not familiar with how the abstraction is constructed around Account

The separation of `validate` and `execute` happened a while ago with some StarkNet version, but here's the documentation on this separation:

https://docs.starknet.io/documentation/architecture_and_concepts/Account_Abstraction/validate_and_execute/

From a very high level, this is such that sequencers can't charge for a failed `validate` but they can for a failed `execute` (not yet though; this will only be chargable with Cairo 1). Given that `validate` is always short and lightweight, sequencers wouldn't be doing huge amount of compensation without getting paid.

**Comment by Eikix** [comment]
> > I'm not familiar with how the abstraction is constructed around Account
> 
> The separation of `validate` and `execute` happened a while ago with some StarkNet version, but here's the documentation on this separation:
> 
> https://docs.starknet.io/documentation/architecture_and_concepts/Account_Abstraction/validate_and_execute/
> 
> From a very high level, this is such that sequencers can't charge for a failed `validate` but they can for a failed `execute` (not yet though; this will only be chargable with Cairo 1). Given that `validate` is always short and lightweight, sequencers wouldn't be doing huge amount of compensation without getting paid.

Oh yeah my apologies, I meant how the abstraction is constructed around the `Account` trait in starknet-rs

**Comment by xJonathanLEI** [comment]
> What I still have trouble understanding is why `add_invoke_transaction` (my understanding is that it submits a transaction to Starknet via JSON RPC) takes as an argument a `BroadcastedInvokeTransaction` and not an `InvokeTransactionRequest`?

First of all, I don't really have much control over the naming of JSON-RPC models. As mentioned, the model code is deterministically [derived](https://github.com/xJonathanLEI/starknet-jsonrpc-codegen) from the official spec, which dictates that `add_invoke_transaction` takes a `BroadcastedInvokeTransaction`.

However, it's wrong to say `JsonRpcClient` doesn't take `InvokeTransactionRequest`. It does when it's acting as `Provider`. So while this won't work (because it's _not_ acting as `Provider`):

```rust
// Not working
rpc_client.add_invoke_transaction(InvokeFunctionTransactionRequest { ... });
```

This would work fine:

```rust
// This works
<JsonRpcClient<HttpTransport> as Provider>::add_transaction(
    &rpc_client,
    TransactionRequest::InvokeFunction(InvokeFunctionTransactionRequest {
        ...
    }),
);
```

Note that you can just write `rpc_client.add_transaction(`. I'm just being explicit that you're using it as that trait.

This complication comes from the fact that, again, the library is currently still sequencer-centric and not modeled after JSON-RPC, because sequencer has been the only feature complete way to interact with the network until very recently. Even today JSON-RPC is still missing some important features compared to the sequencer gateway.

**Comment by xJonathanLEI** [comment]
In general, if you're authoring a function that needs to use a JSON-RPC client, don't take a `JsonRpcClient` type as argument. Instead, take `Provider` as a generic param such that your function works for _all_ `Provider` implementations, including the sequencer gateway.

**Comment by Eikix** [comment]
Ahhh okay! Understood

**Comment by xJonathanLEI** [comment]
> I meant how the abstraction is constructed around the `Account` trait in starknet-rs

Account contract related abstraction is built around the `Account` and `ConnectedAccount` traits. You can check out the traits and the related types (e.g. `Execution`). Another good way to understand how it gets the job done would be to check how the `SingleOwnerAccount` type implements these two traits.
